File: 03/gemini-3.0-pro/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var totalJoltage int64 = 0
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}

		maxJoltage := 0
		found := false

		for d1 := 9; d1 >= 1; d1-- {
			c1 := string(rune('0' + d1))
			idx1 := strings.Index(line, c1)

			if idx1 == -1 {
				continue
			}

			for d2 := 9; d2 >= 1; d2-- {
				c2 := string(rune('0' + d2))
				idx2 := strings.LastIndex(line, c2)

				if idx2 == -1 {
					continue
				}

				if idx1 < idx2 {
					maxJoltage = d1*10 + d2
					found = true
					break
				}
			}
			if found {
				break
			}
		}
		totalJoltage += int64(maxJoltage)
	}

	fmt.Printf("Day 03 Part 1: %d\n", totalJoltage)

	// Reset file pointer for Part 2
	file.Seek(0, 0)
	scanner = bufio.NewScanner(file)
	var totalJoltagePart2 int64 = 0

	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}

		var result strings.Builder
		currentIndex := 0
		digitsNeeded := 12
		length := len(line)

		for i := 0; i < digitsNeeded; i++ {
			remainingNeeded := digitsNeeded - 1 - i
			searchEndIndex := length - 1 - remainingNeeded

			bestDigit := -1
			bestDigitIndex := -1

			for d := 9; d >= 1; d-- {
				c := string(rune('0' + d))
				idx := strings.Index(line[currentIndex:], c)

				if idx != -1 {
					realIdx := currentIndex + idx
					if realIdx <= searchEndIndex {
						bestDigit = d
						bestDigitIndex = realIdx
						break
					}
				}
			}

			if bestDigit != -1 {
				result.WriteString(fmt.Sprintf("%d", bestDigit))
				currentIndex = bestDigitIndex + 1
			}
		}

		if result.Len() == 12 {
			val := int64(0)
			fmt.Sscanf(result.String(), "%d", &val)
			totalJoltagePart2 += val
		}
	}

	fmt.Printf("Day 03 Part 2: %d\n", totalJoltagePart2)
}

File: 03/gemini-3.0-pro/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt")
        .expect("Failed to read input");

    let mut total_joltage: i64 = 0;

    for line in input.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let mut max_joltage = 0;
        let mut found = false;

        for d1 in (1..=9).rev() {
            let c1 = char::from_digit(d1, 10).unwrap();
            let idx1 = line.find(c1);

            if idx1.is_none() {
                continue;
            }
            let idx1 = idx1.unwrap();

            for d2 in (1..=9).rev() {
                let c2 = char::from_digit(d2, 10).unwrap();
                let idx2 = line.rfind(c2);

                if idx2.is_none() {
                    continue;
                }
                let idx2 = idx2.unwrap();

                if idx1 < idx2 {
                    max_joltage = d1 * 10 + d2;
                    found = true;
                    break;
                }
            }
            if found {
                break;
            }
        }
        total_joltage += max_joltage as i64;
    }

    println!("Day 03 Part 1: {}", total_joltage);

    let mut total_joltage_part2: i64 = 0;

    for line in input.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let mut result = String::new();
        let mut current_index = 0;
        let digits_needed = 12;
        let len = line.len();

        for i in 0..digits_needed {
            let remaining_needed = digits_needed - 1 - i;
            let search_end_index = len - 1 - remaining_needed;

            let mut best_digit = -1;
            let mut best_digit_index = -1;

            for d in (1..=9).rev() {
                let c = char::from_digit(d, 10).unwrap();
                // Find first occurrence starting from current_index
                if let Some(idx) = line[current_index..].find(c) {
                    let real_idx = current_index + idx;
                    if real_idx <= search_end_index {
                        best_digit = d as i32;
                        best_digit_index = real_idx as i32;
                        break;
                    }
                }
            }

            if best_digit != -1 {
                result.push_str(&best_digit.to_string());
                current_index = (best_digit_index + 1) as usize;
            }
        }

        if result.len() == 12 {
            total_joltage_part2 += result.parse::<i64>().unwrap();
        }
    }

    println!("Day 03 Part 2: {}", total_joltage_part2);
}
